File: NetNodeClient/src/main/java/com/ethereal/NetNode/blue.java
package com.ethereal.NetNode;

import com.ethereal.client.Client.Abstract.Client;
import com.ethereal.client.Client.Abstract.ClientConfig;
import com.ethereal.client.Client.ClientCore;
import com.ethereal.client.Client.Event.Delegate.OnConnectFailDelegate;
import com.ethereal.client.Client.Event.Delegate.OnDisConnectDelegate;
import com.ethereal.client.Client.WebSocket.WebSocketClient;
import com.ethereal.client.Core.Enums.NetType;
import com.ethereal.client.Core.Model.AbstractTypes;
import com.ethereal.client.Core.Model.TrackException;
import com.ethereal.client.Net.Abstract.Net;
import com.ethereal.client.Net.NetCore;
import com.ethereal.client.Net.NetNode.Model.NetNode;
import com.ethereal.client.Net.NetNode.Request.ServerNetNodeRequest;
import com.ethereal.client.Net.NetNode.Service.ClientNetNodeService;
import com.ethereal.client.Request.Abstract.Request;
import com.ethereal.client.Request.RequestCore;
import com.ethereal.client.Service.Abstract.Service;
import com.ethereal.client.Service.ServiceCore;
import com.ethereal.client.Utils.AutoResetEvent;
import org.javatuples.Pair;

import java.util.concurrent.TimeUnit;

public class blue extends Net {
    public WebSocketNetConfig getConfig() {
        return (WebSocketNetConfig)config;
    }
    AutoResetEvent netNodeSign = new AutoResetEvent(false);


    public blue(String name){
        super(name);
        netType = NetType.WebSocket;
        config = new WebSocketNetConfig();
    }
    @Override
    public boolean publish() throws Exception {
        try {
            //分布式模式
            if(config.getNetNodeMode()){
                //注册数据类型
                AbstractTypes types = new AbstractTypes();
                types.add(Integer.class,"Int");
                types.add(Long.class,"Long");
                types.add(String.class,"String");
                types.add(Boolean.class,"Bool");
                types.add(NetNode.class,"NetNode");
                //注册网关
                Net net = NetCore.register(new WebSocketNet(String.format("NetNode-%s", name)));
                net.getLogEvent().register(this::onLog);
                net.getExceptionEvent().register(this::onException);
                //注册服务
                Service netNodeService = ServiceCore.register(net,new ClientNetNodeService(),"ClientNetNodeService",types);
                //注册请求
                ServerNetNodeRequest netNodeRequest = RequestCore.register(net,ServerNetNodeRequest.class,"ServerNetNodeService",types);
                new Thread(() -> {
                    while (NetCore.get(name)!=null){
                        try {
                            NetNodeSearch();
                        }
                        catch (Exception e) {
                            onException(new TrackException(e));
                        }
                        finally {
                            try {
                                netNodeSign.waitOne(config.getNetNodeHeartInterval(), TimeUnit.MILLISECONDS);
                            } catch (InterruptedException e) {
                                onException(new TrackException(e));
                            }
                        }
                    }
                }).start();
            }
            else {
                for(Request request : requests.values()){
                    request.getClient().connect();
                }
            }
        }
        catch (TrackException e){
            onException(e);
        }
        catch (Exception e){
            onException(new TrackException(e));
        }
        return true;
    }

    private void NetNodeSearch() throws TrackException {
        synchronized (this){
            boolean flag = false;
            for(Request request : requests.values()){
                if(request.getClient() == null){
                    flag = true;
                    break;
                }
            }
            if(flag){
                Client client = null;
                Net net = NetCore.get(String.format("NetNode-%s", name));
                if(net==null)throw new TrackException(TrackException.ErrorCode.Runtime, String.format("NetNode-%s 未找到", name));
                for (Pair<String, ClientConfig> item: config.getNetNodeIps()) {
                    String prefixes = item.getValue0();
                    ClientConfig config = item.getValue1();
                    client = ClientCore.register(net,"ServerNetNodeService",new WebSocketClient(prefixes));
                    if(client == null)throw new TrackException(TrackException.ErrorCode.Runtime, String.format("%s-%s 服务未找到", net.getName(),"ServerNetNodeService"));
                    net.getConfig().setNetNodeMode(false);
                    ((WebSocketClient)client).getConfig().setSyncConnect(true);
                    //启动连接
                    client.connect();
                    try{
                        if(client.isConnect()){
                            Request netNodeRequest = RequestCore.get(net,"ServerNetNodeService");
                            if(netNodeRequest == null)throw new TrackException(TrackException.ErrorCode.Runtime,String.format("%s-%s 查找不到该请求", name,"ServerNetNodeService"));
                            for (Request request : requests.values()) {
                                if(request.getClient() == null){
                                    ServerNetNodeRequest serverNetNodeRequest = (RequestCore.get(net,"ServerNetNodeService"));
                                    //获取服务节点
                                    NetNode node = serverNetNodeRequest.GetNetNode("ServerNetNodeService");

                                    if(node != null){
                                        //注册连接并启动连接
                                        Client requestClient = ClientCore.register(request,new WebSocketClient(node.getPrefixes()[0]));
                                        requestClient.getConnectFailEvent().register(new OnConnectFailDelegate() {
                                            @Override
                                            public void OnConnectFail(Client client) {
                                                client.getConnectFailEvent().unRegister(this);
                                                ClientCore.unregister(client.getNetName(),client.getServiceName());
                                            }
                                        });
                                        requestClient.getDisConnectEvent().register(new OnDisConnectDelegate() {
                                            @Override
                                            public void OnDisConnect(Client client){
                                                try{
                                                    client.getDisConnectEvent().unRegister(this);
                                                    ClientCore.unregister(client.getNetName(),client.getServiceName());
                                                    netNodeSign.set();
                                                }
                                                catch (Exception e){
                                                    client.onException(new TrackException(e));
                                                }
                                            }
                                        });
                                        requestClient.connect();
                                    }
                                    else {
                                        throw new TrackException(TrackException.ErrorCode.Runtime, String.format("%s-%s 在NetNode分布式节点中未找到节点", name,request.getName()));
                                    }
                                }
                            }
                            break;
                        }
                    }
                    catch (TrackException e){
                        onException(e);
                    }
                    catch (Exception e){
                        onException(new TrackException(e));
                    }
                    finally {
                        ClientCore.unregister(net,"ServerNetNodeService");
                    }
                }
            }
        }
    }
}

File: NetNodePluginServer/blue.py
import EtherealC.Service.ServiceCore

from EtherealS.Core.Model.TrackException import TrackException, ExceptionCode
from EtherealS.Net.Abstract.Net import Net
from EtherealS.Net.WebSocket.WebSocketNetConfig import WebSocketNetConfig


class WebSocketNet(Net):
    def __init__(self, name):
        super().__init__(name=name, config=WebSocketNetConfig())
        import threading
        self.connectSign = threading.Event()

    def Publish(self):
        def reactorStart():
            from twisted.internet import reactor
            if not reactor.running:
                reactor.suggestThreadPoolSize(10)
                reactor.run(False)

        import threading
        threading.Thread(target=reactorStart).start()
        self.server.Start()
        if self.config.netNodeMode:
            # 服务端
            from EtherealS.Core.Model.AbstractTypes import AbstractTypes
            from EtherealS.Net.NetNode.Model.NetNode import NetNode
            from EtherealS.Net.NetNode.NetNodeServer.Service.ServerNodeService import ServerNodeService
            from EtherealS.Net.NetNode.NetNodeServer.Request.ClientNodeRequest import ClientNodeRequest
            from EtherealS.Service import ServiceCore
            from EtherealS.Request import RequestCore
            types: AbstractTypes = AbstractTypes()
            types.add(type=int, type_name="Int")
            types.add(type=str, type_name="String")
            types.add(type=bool, type_name="Bool")
            types.add(type=type(NetNode()), type_name="NetNode")
            ServiceCore.Register(net=self, service=ServerNodeService("ServerNetNodeService", types))
            RequestCore.Register(net=self, request=ClientNodeRequest("ClientNetNodeService", types))

            # 客户端
            from EtherealC.Core.Model.AbstractTypes import AbstractTypes
            from EtherealC.Service import ServiceCore
            from EtherealC.Request import RequestCore
            from EtherealC.Net import NetCore
            from EtherealC.Net.WebSocket.WebSocketNet import WebSocketNet
            from EtherealC.Client import ClientCore
            from EtherealC.Client.Abstract import Client
            from EtherealC.Client.WebSocket.WebSocketClient import WebSocketClient
            from EtherealS.Net.NetNode.NetNodeClient.Service.ClientNodeService import ClientNodeService
            from EtherealS.Net.NetNode.NetNodeClient.Request.ServerNodeRequest import ServerNodeRequest
            from EtherealS.Core.Model.TrackException import TrackException, ExceptionCode

            import threading

            def NetNodeSearchRunner():
                from EtherealS.Net import NetCore as ServerNetCore
                while ServerNetCore.Get(self.name) is not None:
                    try:
                        for tuple in self.config.netNodeIps:
                            prefixes = tuple["prefixes"]
                            clientConfig = tuple["config"]
                            net = NetCore.Get("NetNodeClient-{0}".format(prefixes))
                            if net is None:
                                net = NetCore.Register(WebSocketNet(name="NetNodeClient-{0}".format(prefixes)))
                                net.config.netNodeMode = False
                                net.log_event.register(self.OnLog)
                                net.exception_event.register(self.OnException)
                            serverNodeRequest = EtherealC.Request.RequestCore.Get(net=net, service_name="ServerNetNodeService")
                            if serverNodeRequest is None:
                                serverNodeRequest = EtherealC.Request.RequestCore.Register(net=net, request=ServerNodeRequest())
                            clientNodeService = EtherealC.Service.ServiceCore.Get(net=net, service_name="ClientNetNodeService")
                            if clientNodeService is None:
                                clientNodeService: ClientNodeService = EtherealC.Service.ServiceCore.Register(net=net, service=ClientNodeService())

                            if serverNodeRequest.client is None:
                                client: Client = ClientCore.Register(request=serverNodeRequest,client=WebSocketClient(prefixes))
                                if clientConfig is not None:
                                    client.config = clientConfig
                                client.connectSuccess_event.register(self.ClientNodeConnectSuccess)
                                client.connectFail_event.register(self.ClientNodeConnectFail)
                                client.disconnect_event.register(self.ClientNodeDisConnect)
                                net.Publish()
                    except Exception as e:
                        self.OnException(TrackException(code=ExceptionCode.Runtime, message=e.args, exception=e))
                    finally:
                        self.connectSign.wait(self.config.netNodeHeartbeatCycle / 1000)
                        self.connectSign.clear()

            threading.Thread(target=NetNodeSearchRunner).start()
        return True

    def ClientNodeConnectSuccess(self, client):
        from EtherealC.Request import RequestCore
        from EtherealS.Net.NetNode.NetNodeClient.Request.ServerNodeRequest import ServerNodeRequest
        serverNodeRequest: ServerNodeRequest = RequestCore.Get(net_name="NetNodeClient-{0}".format(client.prefixes),
                                                                service_name="ServerNetNodeService")
        if serverNodeRequest is not None:
            from EtherealS.Net.NetNode.Model.NetNode import NetNode
            node = NetNode()
            node.Prefixes = self.server.prefixes
            node.Name = self.name
            for service in self.services.values():
                from EtherealS.Net.NetNode.Model.ServiceNode import ServiceNode
                serviceNode = ServiceNode()
                serviceNode.name = service.name
                node.Services[serviceNode.name] = serviceNode
            for request in self.requests.values():
                from EtherealS.Net.NetNode.Model.RequestNode import RequestNode
                requestNode = RequestNode()
                requestNode.name = request.name
                node.Requests[requestNode.name] = requestNode
            serverNodeRequest.Register(node)
        else:
            raise TrackException(code=ExceptionCode.Runtime,
                                 message="EtherealC中未找到 NetNodeClient-{0}-ServerNodeService".format(client.prefixes))

    def ClientNodeConnectFail(self, client):
        from EtherealC.Client import ClientCore
        ClientCore.UnRegister(net_name=client.net_name, service_name=client.service_name)

    def ClientNodeDisConnect(self, client):
        from EtherealC.Client import ClientCore
        self.connectSign.set()
        ClientCore.UnRegister(net_name=client.net_name, service_name=client.service_name)


